2_process/src/aggregate_observations.R

```diff
@@ -0,0 +1,150 @@
+  #' whether segment summaries should defer to NWIS observations when available. Defaults 
+  #' to FALSE. If TRUE, any observations coming from a non-NWIS site will be omitted.
+  #' 
+  #' @value returns a data frame containing 1 mean/min/max value for each unique segment-date
+  #' 
+  
+  # Format water quality data to include segment assignments
+  vars <- c("site_id", "lon", "lat", "datum", "subsegid",
+            "bird_dist_to_subseg_m", "segidnat")
+  sites_w_segs <- sites_w_segs %>%
+    select(any_of(vars))
+  
+  obs_data_w_segs <- combine_wq_data(wqp_data, nwis_data) %>%
+    filter(site_no %in% sites_w_segs$site_id) %>%
+    left_join(.,sites_w_segs,
+              by=c("site_no" = "site_id")) 
+  
+  # Aggregate observations 
+  if(aggr_method == "reach"){
+    
+    obs_data_aggr <- obs_data_w_segs %>%
+      group_by(subsegid, Date) %>%
+      # If prefer_nwis_sites = TRUE, check whether data for that segment comes from
+      # multiple sources (i.e., WQP and NWIS). If multiple sources, filter out any
+      # non-NWIS samples for that segment-date; otherwise, retain all samples
+      {if(prefer_nwis_sites == "TRUE"){
+        filter(.,if(n_distinct(dat_src) > 1) dat_src == "NWIS" else TRUE)
+      } else {.}
+      } %>%
+      summarize(mean_value = round(mean(Value, na.rm = TRUE), 1),
+                min_value = min(c(mean_value,Value_Min), na.rm = TRUE),
+                max_value = max(c(mean_value,Value_Max), na.rm = TRUE),
+                sd_value = round(sd(Value, na.rm = TRUE), 2),
+                n_value = length(!is.na(Value)),
+                site_ids = paste0(unique(site_no), collapse = ","),
+                .groups = "keep") %>%
+      # suppress any warnings related to taking min/max summaries
+      # when no non-NA data exist for a reach-date; in these cases, 
+      # Inf will be returned for min_value, max_value
+      suppressWarnings() %>%
+      mutate(cv_value = (sd_value/mean_value)) %>%
+      ungroup()
+    
+  } else {
+    
+    # For aggr_method = "site", group by lat/lon since in the WQP database, there
+    # may exist different site names that share the same geographic coordinates
+    obs_data_aggr <- obs_data_w_segs %>%
+      group_by(lat,lon,Date) %>%
+      summarize(mean_value = round(mean(Value, na.rm = TRUE), 1),
+                min_value = min(c(mean_value,Value_Min), na.rm = TRUE),
+                max_value = max(c(mean_value,Value_Max), na.rm = TRUE),
+                sd_value = round(sd(Value, na.rm = TRUE), 2),
+                n_value = length(!is.na(Value)),
+                site_ids = paste0(unique(site_no), collapse = ","),
+                .groups = "keep") %>%
+      # suppress any warnings related to taking min/max summaries
+      # when no non-NA data exist for a reach-date; in these cases, 
+      # Inf will be returned for min_value, max_value
+      suppressWarnings() %>%
+      mutate(cv_value = (sd_value/mean_value)) %>%
+      ungroup()
+    
+  }
```

Comment:
Could you do just split on the grouping but have the same `summarise` step those lines are the same? Warning: I haven't tried this and I'm a `dplyr` newbie.
```suggestion
    grouped_obs <- obs_data_w_segs %>%
      group_by(subsegid, Date) %>%
      # If prefer_nwis_sites = TRUE, check whether data for that segment comes from
      # multiple sources (i.e., WQP and NWIS). If multiple sources, filter out any
      # non-NWIS samples for that segment-date; otherwise, retain all samples
      {if(prefer_nwis_sites == "TRUE"){
        filter(.,if(n_distinct(dat_src) > 1) dat_src == "NWIS" else TRUE)
      } else {.}
  } else {
    
    # For aggr_method = "site", group by lat/lon since in the WQP database, there
    # may exist different site names that share the same geographic coordinates
    grouped_obs <- obs_data_w_segs %>%
      group_by(lat,lon,Date) 
  }
  
  obs_data_aggr <- grouped_obs %>%
     summarize(mean_value = round(mean(Value, na.rm = TRUE), 1),
                min_value = min(c(mean_value,Value_Min), na.rm = TRUE),
                max_value = max(c(mean_value,Value_Max), na.rm = TRUE),
                sd_value = round(sd(Value, na.rm = TRUE), 2),
                n_value = length(!is.na(Value)),
                site_ids = paste0(unique(site_no), collapse = ","),
                .groups = "keep") %>%
      # suppress any warnings related to taking min/max summaries
      # when no non-NA data exist for a reach-date; in these cases, 
      # Inf will be returned for min_value, max_value
      suppressWarnings() %>%
      mutate(cv_value = (sd_value/mean_value)) %>%
      ungroup()
```